base/ltshipout.dtx

```diff
@@ -636,17 +730,67 @@
 \cs_new:Npn \@@_execute_test_level: {
   \int_compare:nNnT
      \l_@@_group_level_tl < \tex_currentgrouplevel:D 
-     \tex_aftergroup:D
-  \@@_execute_cont:
+     \tex_aftergroup:D \@@_execute_cont:
 }
 %    \end{macrocode}
 %  \end{macro}
 %
 %
 %  \begin{macro}{\@@_execute_cont:}
+%    This does the actual shipout running several hooks as part of it.
+%    The code for them is passed as argument \verb=#2= to \verb=#4=
+%    to \cs{@@_execute_main_cont:Nnnn}; the first argument is the box
+%    to be shipped out.
+%    \begin{macrocode}
+\cs_new:Npn \@@_execute_cont: {
+  \@@_execute_main_cont:Nnnn
+     \l_shipout_box
+     { \hook_use:n {shipout/before} }
+     { \hook_if_empty:nF {shipout/foreground}
+         { \@@_add_foreground_picture:n
+             { \hook_use:n {shipout/foreground} } }
+%    \end{macrocode}
+%    If the user hook for the background (\hook{shipout/background}) has
+%    no code, there might still code in the kernel hook so we need to
+%    test for this too.
+%    We only test for the \cs{@kernel@before@shipout@background}
+%    though. If the  \cs{@kernel@after@shipout@background} needs
+%    executing even if the user hook is empty then we can add another
+%    test (or the kernel could put something into the before hook).
+%                   
+% \changes{v1.0d}{2020/11/23}{Check for both kernel and user hook (gh/431)}
+% \changes{v1.0f}{2021/01/08}{Added another kernel hook for more
+%    flexibility (cf \texttt{https://github.com/pgf-tikz/pgf/issues/960}}
+%    \begin{macrocode}
+       \bool_lazy_and:nnF
+         { \hook_if_empty_p:n {shipout/background} }
+         { \tl_if_empty_p:N \@kernel@before@shipout@background }
+         { \@@_add_background_picture:n
+           { \@kernel@before@shipout@background
+             \hook_use:n {shipout/background}
+             \@kernel@after@shipout@background }
+         }
```

Comment:
so it is and so it is documented. My expectation is that we don't have a need for just filling the after kernel hook. So I wanted to omit that test for every page (maybe a dumb optimization but I'm an old timer remember (have seen 7bit machines :-)

Reply:
> so it is and so it is documented

Oh, I missed that, sorry.

> maybe a dumb optimization

Not at all.  I was going to suggest reversing the tests if we are more likely to use the kernel hooks :) (the test short-circuits)

But (I think) this hook was added with PGF in mind (@u-fischer knows better), so if they add something to `\@kernel@after@shipout@background` and nothing else is in the other hook, their code won't run... But then they can just add `\@empty` to the before hook, so I'm probably complaining on something irrelevant

Reply:
I saw this and added some to the before hook.

Reply:
wait a second are you saying that PGF is adding to a kernel internal hook?

Reply:
I _think_ it's from [this issue](https://github.com/pgf-tikz/pgf/issues/960), that requires PGF to insert code in the `shipout/background` hook, after the `top-level` label (which is no longer possible within that hook).  I think this is temporary while the pdfresources are not done, but I'm not completely sure

Reply:
@FrankMittelbach  No I don't think that PGF is using the kernel hook (which they can't anyway as it isn't there yet ;-)). But the discussion in the pgf issue triggered my request for the new `\@kernel@after@shipout@background` for my code. I'm not quite sure what to do about the pgf problem.   When  the pdf resource management is there it will be resolved, as my code is in the right place in the new hook, but until then they need some work around. Probably they should simply add the resource earlier, even if it is not needed in the end, on the whole it is only one pdf object.

Reply:
ok panic other then :-)